File: go/stale_memory.go
package brain

import (
	"fmt"
	"sort"
	"strings"
)

var DefaultStaleMemoryWeights = StaleMemoryWeights{
	Recency:         0.4,
	AccessFrequency: 0.25,
	RetentionValue:  0.35,
}

type StaleMemoryWeights struct {
	Recency         float64 `json:"recency"`
	AccessFrequency float64 `json:"accessFrequency"`
	RetentionValue  float64 `json:"retentionValue"`
}

type StaleMemoryInput struct {
	MemoryID       string         `json:"memoryId"`
	CreatedAt      string         `json:"createdAt"`
	LastAccessedAt string         `json:"lastAccessedAt"`
	AccessCount    int            `json:"accessCount"`
	RetentionValue float64        `json:"retentionValue"`
	Metadata       map[string]any `json:"metadata,omitempty"`
}

type StaleMemoryBreakdown struct {
	Recency         float64 `json:"recency"`
	AccessFrequency float64 `json:"accessFrequency"`
	RetentionValue  float64 `json:"retentionValue"`
}

type StaleMemoryEvaluation struct {
	MemoryID       string               `json:"memoryId"`
	Metadata       map[string]any       `json:"metadata,omitempty"`
	CreatedAt      string               `json:"createdAt"`
	LastAccessedAt string               `json:"lastAccessedAt"`
	AccessCount    int                  `json:"accessCount"`
	RetentionValue float64              `json:"retentionValue"`
	RecencyMS      int64                `json:"recencyMs"`
	StaleScore     float64              `json:"staleScore"`
	StaleCandidate bool                 `json:"staleCandidate"`
	Reasons        []string             `json:"reasons"`
	Breakdown      StaleMemoryBreakdown `json:"breakdown"`
}

type StaleMemoryEvaluationOptions struct {
	Now                      string
	MinimumRecencyMS         int64
	RecencyHorizonMS         int64
	AccessFrequencyCapPerDay float64
	StaleThreshold           float64
	Weights                  StaleMemoryWeights
	Memories                 []StaleMemoryInput
}

type StaleMemoryEvaluationResult struct {
	EvaluatedAt              string                  `json:"evaluatedAt"`
	EvaluationMode           string                  `json:"evaluationMode"`
	MinimumRecencyMS         int64                   `json:"minimumRecencyMs"`
	RecencyHorizonMS         int64                   `json:"recencyHorizonMs"`
	AccessFrequencyCapPerDay float64                 `json:"accessFrequencyCapPerDay"`
	StaleThreshold           float64                 `json:"staleThreshold"`
	Weights                  StaleMemoryWeights      `json:"weights"`
	ScoredMemories           []StaleMemoryEvaluation `json:"scoredMemories"`
	StaleCandidates          []StaleMemoryEvaluation `json:"staleCandidates"`
	StaleCandidateCount      int                     `json:"staleCandidateCount"`
}

type StaleMemoryMaskingDecision struct {
	MemoryID             string                         `json:"memoryId"`
	MemoryKind           string                         `json:"memoryKind,omitempty"`
	InactiveForRetrieval bool                           `json:"inactiveForRetrieval"`
	Masking              YoungGenerationMaskingMetadata `json:"masking"`
	SourceEvaluation     StaleMemoryEvaluation          `json:"sourceEvaluation"`
}

type StaleMemoryMaskingDecisionOptions struct {
	Evaluation StaleMemoryEvaluationResult
	MaskedAt   string
	MaskedBy   string
	Reason     string
	Provenance map[string]any
}

type StaleMemoryMaskingDecisionResult struct {
	DecisionMode           string                       `json:"decisionMode"`
	MaskedAt               string                       `json:"maskedAt"`
	MaskedBy               string                       `json:"maskedBy"`
	Reason                 string                       `json:"reason"`
	MaskedDecisions        []StaleMemoryMaskingDecision `json:"maskedDecisions"`
	MaskedDecisionCount    int                          `json:"maskedDecisionCount"`
	DeferredCandidates     []StaleMemoryEvaluation      `json:"deferredCandidates"`
	DeferredCandidateCount int                          `json:"deferredCandidateCount"`
}

func EvaluateStaleMemories(options StaleMemoryEvaluationOptions) (StaleMemoryEvaluationResult, error) {
	now, err := parseRequiredTime(options.Now, "now")
	if err != nil {
		return StaleMemoryEvaluationResult{}, err
	}
	if options.MinimumRecencyMS <= 0 {
		options.MinimumRecencyMS = 7 * 24 * 60 * 60 * 1000
	}
	if options.RecencyHorizonMS <= 0 {
		options.RecencyHorizonMS = 30 * 24 * 60 * 60 * 1000
	}
	if options.AccessFrequencyCapPerDay <= 0 {
		options.AccessFrequencyCapPerDay = 4
	}
	if options.StaleThreshold <= 0 {
		options.StaleThreshold = 0.65
	}
	if options.Weights == (StaleMemoryWeights{}) {
		options.Weights = DefaultStaleMemoryWeights
	}
	result := StaleMemoryEvaluationResult{
		EvaluatedAt:              now.UTC().Format(timeFormat()),
		EvaluationMode:           "offline-suggestion-only",
		MinimumRecencyMS:         options.MinimumRecencyMS,
		RecencyHorizonMS:         options.RecencyHorizonMS,
		AccessFrequencyCapPerDay: options.AccessFrequencyCapPerDay,
		StaleThreshold:           options.StaleThreshold,
		Weights:                  options.Weights,
		ScoredMemories:           []StaleMemoryEvaluation{},
		StaleCandidates:          []StaleMemoryEvaluation{},
	}
	for _, memory := range options.Memories {
		createdAt, err := parseRequiredTime(memory.CreatedAt, "createdAt")
		if err != nil {
			return StaleMemoryEvaluationResult{}, err
		}
		lastAccessedAt, err := parseRequiredTime(memory.LastAccessedAt, "lastAccessedAt")
		if err != nil {
			return StaleMemoryEvaluationResult{}, err
		}
		if lastAccessedAt.Before(createdAt) {
			return StaleMemoryEvaluationResult{}, fmt.Errorf("lastAccessedAt must not be earlier than createdAt")
		}
		recencyMS := now.Sub(lastAccessedAt).Milliseconds()
		recencyScore := clamp01(float64(recencyMS) / float64(options.RecencyHorizonMS))
		ageDays := maxFloat64(float64(now.Sub(createdAt).Hours()/24), 1)
		accessPerDay := float64(memory.AccessCount) / ageDays
		accessScore := clamp01(1 - (accessPerDay / options.AccessFrequencyCapPerDay))
		retentionScore := clamp01(1 - memory.RetentionValue)
		staleScore := round4(
			recencyScore*options.Weights.Recency +
				accessScore*options.Weights.AccessFrequency +
				retentionScore*options.Weights.RetentionValue,
		)
		reasons := []string{}
		if recencyScore >= 0.5 {
			reasons = append(reasons, "stale-recency")
		}
		if accessScore >= 0.5 {
			reasons = append(reasons, "low-access-frequency")
		}
		if retentionScore >= 0.5 {
			reasons = append(reasons, "low-retention-value")
		}
		evaluation := StaleMemoryEvaluation{
			MemoryID:       memory.MemoryID,
			Metadata:       copyMap(memory.Metadata),
			CreatedAt:      mustNormalizeTimeString(memory.CreatedAt),
			LastAccessedAt: mustNormalizeTimeString(memory.LastAccessedAt),
			AccessCount:    memory.AccessCount,
			RetentionValue: clamp01(memory.RetentionValue),
			RecencyMS:      recencyMS,
			StaleScore:     staleScore,
			StaleCandidate: recencyMS >= options.MinimumRecencyMS && staleScore >= options.StaleThreshold,
			Reasons:        reasons,
			Breakdown: StaleMemoryBreakdown{
				Recency:         round4(recencyScore),
				AccessFrequency: round4(accessScore),
				RetentionValue:  round4(retentionScore),
			},
		}
		result.ScoredMemories = append(result.ScoredMemories, evaluation)
		if evaluation.StaleCandidate {
			result.StaleCandidates = append(result.StaleCandidates, evaluation)
		}
	}
	sort.SliceStable(result.ScoredMemories, func(i, j int) bool {
		return result.ScoredMemories[i].StaleScore > result.ScoredMemories[j].StaleScore
	})
	sort.SliceStable(result.StaleCandidates, func(i, j int) bool {
		return result.StaleCandidates[i].StaleScore > result.StaleCandidates[j].StaleScore
	})
	result.StaleCandidateCount = len(result.StaleCandidates)
	return result, nil
}

func CreateStaleMemoryMaskingDecisions(options StaleMemoryMaskingDecisionOptions) StaleMemoryMaskingDecisionResult {
	maskedAt := options.MaskedAt
	if strings.TrimSpace(maskedAt) == "" {
		maskedAt = options.Evaluation.EvaluatedAt
	}
	maskedAt = mustNormalizeTimeString(maskedAt)
	maskedBy := coalesceString(options.MaskedBy, "offline-consolidation")
	reason := coalesceString(options.Reason, "stale-low-value")
	result := StaleMemoryMaskingDecisionResult{
		DecisionMode:       "offline-suggestion-only",
		MaskedAt:           maskedAt,
		MaskedBy:           maskedBy,
		Reason:             reason,
		MaskedDecisions:    []StaleMemoryMaskingDecision{},
		DeferredCandidates: []StaleMemoryEvaluation{},
	}
	for _, candidate := range options.Evaluation.StaleCandidates {
		memoryKind, _ := candidate.Metadata["memoryKind"].(string)
		if candidate.RetentionValue <= 0.5 {
			result.MaskedDecisions = append(result.MaskedDecisions, StaleMemoryMaskingDecision{
				MemoryID:             candidate.MemoryID,
				MemoryKind:           memoryKind,
				InactiveForRetrieval: true,
				Masking: YoungGenerationMaskingMetadata{
					IsMasked:      true,
					MaskedAt:      maskedAt,
					MaskUpdatedAt: maskedAt,
					MaskedBy:      maskedBy,
					Reason:        reason,
					Provenance: mergeMaps(options.Provenance, map[string]any{
						"sourceEvaluationAt": options.Evaluation.EvaluatedAt,
					}),
					Audit: &YoungGenerationMaskingAuditMetadata{
						RuntimePhase:         stringFromMap(options.Provenance, "runtimePhase"),
						RecordedAt:           maskedAt,
						Actor:                maskedBy,
						SourceEvaluationMode: options.Evaluation.EvaluationMode,
					},
				},
				SourceEvaluation: candidate,
			})
		} else {
			result.DeferredCandidates = append(result.DeferredCandidates, candidate)
		}
	}
	result.MaskedDecisionCount = len(result.MaskedDecisions)
	result.DeferredCandidateCount = len(result.DeferredCandidates)
	return result
}

func mergeMaps(values ...map[string]any) map[string]any {
	result := map[string]any{}
	for _, value := range values {
		for key, entry := range value {
			result[key] = entry
		}
	}
	return result
}

func stringFromMap(value map[string]any, key string) string {
	raw, _ := value[key].(string)
	return raw
}

func maxFloat64(left, right float64) float64 {
	if left > right {
		return left
	}
	return right
}

func timeFormat() string {
	return "2006-01-02T15:04:05Z"
}

File: src/stale-memory.js
const DAY_IN_MS = 24 * 60 * 60 * 1000;

const DEFAULT_MINIMUM_RECENCY_MS = DAY_IN_MS;
const DEFAULT_STALE_THRESHOLD = 0.65;
const DEFAULT_MASKED_BY = "offline-consolidation";
const DEFAULT_MASKING_REASON = "stale-low-value";
const DEFAULT_MASKABLE_REASONS = Object.freeze(["low-retention-value"]);
const YOUNG_GENERATION_MEMORY_KINDS = Object.freeze([
  "working_memory",
  "short_term_memory",
]);

const freezeDeep = (value) => {
  if (!value || typeof value !== "object" || Object.isFrozen(value)) {
    return value;
  }

  if (Array.isArray(value)) {
    value.forEach(freezeDeep);
    return Object.freeze(value);
  }

  Object.values(value).forEach(freezeDeep);
  return Object.freeze(value);
};

const cloneObject = (value) => {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return {};
  }

  return { ...value };
};

const roundScore = (value) => Number(value.toFixed(6));

const clampUnitInterval = (value) => Math.min(Math.max(value, 0), 1);

const normalizeNonEmptyString = (value, label) => {
  if (typeof value !== "string") {
    throw new TypeError(`${label} must be a string`);
  }

  const normalizedValue = value.trim();

  if (!normalizedValue) {
    throw new TypeError(`${label} must not be empty`);
  }

  return normalizedValue;
};

const normalizeNonNegativeNumber = (value, label) => {
  if (typeof value !== "number" || Number.isNaN(value) || value < 0) {
    throw new TypeError(`${label} must be a non-negative number`);
  }

  return value;
};

const normalizePositiveNumber = (value, label) => {
  if (typeof value !== "number" || Number.isNaN(value) || value <= 0) {
    throw new TypeError(`${label} must be a positive number`);
  }

  return value;
};

const normalizeUnitInterval = (value, label) => {
  const normalizedValue = normalizeNonNegativeNumber(value, label);

  if (normalizedValue > 1) {
    throw new TypeError(`${label} must be between 0 and 1`);
  }

  return normalizedValue;
};

const normalizeNonEmptyStringArray = (value, label) => {
  if (!Array.isArray(value)) {
    throw new TypeError(`${label} must be an array`);
  }

  return freezeDeep(value.map((entry, index) => normalizeNonEmptyString(entry, `${label}[${index}]`)));
};

const normalizeTimestampMs = (value, label) => {
  const timestamp =
    value instanceof Date ? value.getTime() : new Date(value).getTime();

  if (Number.isNaN(timestamp)) {
    throw new TypeError(`${label} must be a valid date-like value`);
  }

  return timestamp;
};

const toIsoTimestamp = (timestampMs) => new Date(timestampMs).toISOString();

export const DEFAULT_STALE_MEMORY_WEIGHTS = freezeDeep({
  recency: 0.4,
  accessFrequency: 0.25,
  retentionValue: 0.35,
});

const normalizeWeights = (weights = {}) => {
  const mergedWeights = {
    ...DEFAULT_STALE_MEMORY_WEIGHTS,
    ...(weights ?? {}),
  };

  const recency = normalizeNonNegativeNumber(mergedWeights.recency, "weights.recency");
  const accessFrequency = normalizeNonNegativeNumber(
    mergedWeights.accessFrequency,
    "weights.accessFrequency",
  );
  const retentionValue = normalizeNonNegativeNumber(
    mergedWeights.retentionValue,
    "weights.retentionValue",
  );
  const totalWeight = recency + accessFrequency + retentionValue;

  if (totalWeight <= 0) {
    throw new TypeError("weights must contain at least one positive value");
  }

  return freezeDeep({
    recency: roundScore(recency / totalWeight),
    accessFrequency: roundScore(accessFrequency / totalWeight),
    retentionValue: roundScore(retentionValue / totalWeight),
  });
};

const normalizeMemory = (memory, nowMs) => {
  if (!memory || typeof memory !== "object" || Array.isArray(memory)) {
    throw new TypeError("memory entries must be objects");
  }

  const createdAtMs = normalizeTimestampMs(memory.createdAt, "memory createdAt");
  const lastAccessedAtMs =
    memory.lastAccessedAt === undefined || memory.lastAccessedAt === null
      ? createdAtMs
      : normalizeTimestampMs(memory.lastAccessedAt, "memory lastAccessedAt");

  if (lastAccessedAtMs < createdAtMs) {
    throw new TypeError("memory lastAccessedAt must not be earlier than createdAt");
  }

  if (createdAtMs > nowMs) {
    throw new TypeError("memory createdAt must not be in the future");
  }

  if (lastAccessedAtMs > nowMs) {
    throw new TypeError("memory lastAccessedAt must not be in the future");
  }

  return {
    memoryId: normalizeNonEmptyString(memory.memoryId, "memoryId"),
    createdAt: toIsoTimestamp(createdAtMs),
    createdAtMs,
    lastAccessedAt: toIsoTimestamp(lastAccessedAtMs),
    lastAccessedAtMs,
    accessCount: normalizeNonNegativeNumber(
      memory.accessCount ?? 0,
      "memory accessCount",
    ),
    retentionValue: normalizeUnitInterval(
      memory.retentionValue,
      "memory retentionValue",
    ),
    metadata:
      memory.metadata === undefined || memory.metadata === null
        ? null
        : cloneObject(memory.metadata),
  };
};

const compareScoredMemories = (left, right) => {
  if (right.staleScore !== left.staleScore) {
    return right.staleScore - left.staleScore;
  }

  return left.memoryId.localeCompare(right.memoryId);
};

const buildReasons = (breakdown) => {
  const reasons = [];

  if (breakdown.recency >= 0.66) {
    reasons.push("stale-recency");
  }

  if (breakdown.accessFrequency >= 0.66) {
    reasons.push("low-access-frequency");
  }

  if (breakdown.retentionValue >= 0.66) {
    reasons.push("low-retention-value");
  }

  return reasons;
};

const normalizeBreakdown = (value, label) => {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError(`${label} must be an object`);
  }

  return freezeDeep({
    recency: normalizeUnitInterval(value.recency, `${label}.recency`),
    accessFrequency: normalizeUnitInterval(
      value.accessFrequency,
      `${label}.accessFrequency`,
    ),
    retentionValue: normalizeUnitInterval(
      value.retentionValue,
      `${label}.retentionValue`,
    ),
  });
};

const normalizeScoredMemory = (memory, label) => {
  if (!memory || typeof memory !== "object" || Array.isArray(memory)) {
    throw new TypeError(`${label} must be an object`);
  }

  return freezeDeep({
    memoryId: normalizeNonEmptyString(memory.memoryId, `${label}.memoryId`),
    createdAt: toIsoTimestamp(normalizeTimestampMs(memory.createdAt, `${label}.createdAt`)),
    lastAccessedAt: toIsoTimestamp(
      normalizeTimestampMs(memory.lastAccessedAt, `${label}.lastAccessedAt`),
    ),
    ageMs: normalizeNonNegativeNumber(memory.ageMs, `${label}.ageMs`),
    recencyMs: normalizeNonNegativeNumber(memory.recencyMs, `${label}.recencyMs`),
    accessCount: normalizeNonNegativeNumber(memory.accessCount, `${label}.accessCount`),
    accessFrequencyPerDay: normalizeNonNegativeNumber(
      memory.accessFrequencyPerDay,
      `${label}.accessFrequencyPerDay`,
    ),
    retentionValue: normalizeUnitInterval(
      memory.retentionValue,
      `${label}.retentionValue`,
    ),
    staleScore: normalizeUnitInterval(memory.staleScore, `${label}.staleScore`),
    staleCandidate: Boolean(memory.staleCandidate),
    breakdown: normalizeBreakdown(memory.breakdown, `${label}.breakdown`),
    reasons: normalizeNonEmptyStringArray(memory.reasons ?? [], `${label}.reasons`),
    metadata:
      memory.metadata === undefined || memory.metadata === null
        ? null
        : cloneObject(memory.metadata),
  });
};

const normalizeYoungGenerationMemoryKind = (metadata) => {
  if (!metadata || typeof metadata !== "object" || Array.isArray(metadata)) {
    return null;
  }

  return YOUNG_GENERATION_MEMORY_KINDS.includes(metadata.memoryKind)
    ? metadata.memoryKind
    : null;
};

export const evaluateStaleMemories = (options = {}) => {
  const nowMs =
    options.now === undefined ? Date.now() : normalizeTimestampMs(options.now, "now");
  const staleThreshold = normalizeUnitInterval(
    options.staleThreshold ?? DEFAULT_STALE_THRESHOLD,
    "staleThreshold",
  );
  const minimumRecencyMs = normalizeNonNegativeNumber(
    options.minimumRecencyMs ?? DEFAULT_MINIMUM_RECENCY_MS,
    "minimumRecencyMs",
  );
  const weights = normalizeWeights(options.weights);
  const memoryInputs = options.memories ?? [];

  if (!Array.isArray(memoryInputs)) {
    throw new TypeError("memories must be an array");
  }

  const normalizedMemories = memoryInputs.map((memory) =>
    normalizeMemory(memory, nowMs),
  );
  const recencyValues = normalizedMemories.map(
    (memory) => nowMs - memory.lastAccessedAtMs,
  );
  const recencyHorizonMs =
    options.recencyHorizonMs === undefined || options.recencyHorizonMs === null
      ? Math.max(minimumRecencyMs, DAY_IN_MS, ...recencyValues)
      : normalizePositiveNumber(options.recencyHorizonMs, "recencyHorizonMs");
  const accessFrequenciesPerDay = normalizedMemories.map((memory) => {
    const ageInDays = Math.max((nowMs - memory.createdAtMs) / DAY_IN_MS, 1);
    return memory.accessCount / ageInDays;
  });
  const accessFrequencyCapPerDay =
    options.accessFrequencyCapPerDay === undefined ||
    options.accessFrequencyCapPerDay === null
      ? Math.max(1, ...accessFrequenciesPerDay)
      : normalizePositiveNumber(
          options.accessFrequencyCapPerDay,
          "accessFrequencyCapPerDay",
        );
  const evaluatedAt = toIsoTimestamp(nowMs);

  const scoredMemories = normalizedMemories
    .map((memory, index) => {
      const ageMs = nowMs - memory.createdAtMs;
      const recencyMs = nowMs - memory.lastAccessedAtMs;
      const accessFrequencyPerDay = accessFrequenciesPerDay[index];
      const breakdown = freezeDeep({
        recency: roundScore(clampUnitInterval(recencyMs / recencyHorizonMs)),
        accessFrequency: roundScore(
          1 - clampUnitInterval(accessFrequencyPerDay / accessFrequencyCapPerDay),
        ),
        retentionValue: roundScore(1 - memory.retentionValue),
      });
      const staleScore = roundScore(
        breakdown.recency * weights.recency +
          breakdown.accessFrequency * weights.accessFrequency +
          breakdown.retentionValue * weights.retentionValue,
      );
      const staleCandidate =
        recencyMs >= minimumRecencyMs && staleScore >= staleThreshold;

      return freezeDeep({
        memoryId: memory.memoryId,
        createdAt: memory.createdAt,
        lastAccessedAt: memory.lastAccessedAt,
        ageMs,
        recencyMs,
        accessCount: memory.accessCount,
        accessFrequencyPerDay: roundScore(accessFrequencyPerDay),
        retentionValue: memory.retentionValue,
        staleScore,
        staleCandidate,
        breakdown,
        reasons: buildReasons(breakdown),
        metadata: memory.metadata,
      });
    })
    .sort(compareScoredMemories);
  const staleCandidates = scoredMemories.filter((memory) => memory.staleCandidate);

  return freezeDeep({
    evaluatedAt,
    staleThreshold,
    minimumRecencyMs,
    recencyHorizonMs,
    accessFrequencyCapPerDay: roundScore(accessFrequencyCapPerDay),
    weights,
    scoredMemories,
    staleCandidates,
    staleCandidateCount: staleCandidates.length,
    evaluationMode: "offline-suggestion-only",
  });
};

export const createStaleMemoryMaskingDecisions = (options = {}) => {
  const normalizedOptions =
    options && typeof options === "object" && !Array.isArray(options) ? options : {};
  const evaluation = normalizedOptions.evaluation;

  if (!evaluation || typeof evaluation !== "object" || Array.isArray(evaluation)) {
    throw new TypeError("evaluation must be an object");
  }

  if (!Array.isArray(evaluation.staleCandidates)) {
    throw new TypeError("evaluation.staleCandidates must be an array");
  }

  const sourceEvaluationAt = toIsoTimestamp(
    normalizeTimestampMs(evaluation.evaluatedAt, "evaluation.evaluatedAt"),
  );
  const sourceEvaluationMode = normalizeNonEmptyString(
    evaluation.evaluationMode,
    "evaluation.evaluationMode",
  );
  const maskedAt = toIsoTimestamp(
    normalizedOptions.maskedAt === undefined
      ? normalizeTimestampMs(sourceEvaluationAt, "evaluation.evaluatedAt")
      : normalizeTimestampMs(normalizedOptions.maskedAt, "maskedAt"),
  );
  const maskedBy = normalizeNonEmptyString(
    normalizedOptions.maskedBy ?? DEFAULT_MASKED_BY,
    "maskedBy",
  );
  const reason = normalizeNonEmptyString(
    normalizedOptions.reason ?? DEFAULT_MASKING_REASON,
    "reason",
  );
  const maskableReasons = normalizeNonEmptyStringArray(
    normalizedOptions.maskableReasons ?? DEFAULT_MASKABLE_REASONS,
    "maskableReasons",
  );
  const provenance =
    normalizedOptions.provenance === undefined || normalizedOptions.provenance === null
      ? null
      : cloneObject(normalizedOptions.provenance);
  const staleCandidates = evaluation.staleCandidates
    .map((memory, index) =>
      normalizeScoredMemory(memory, `evaluation.staleCandidates[${index}]`),
    )
    .filter((memory) => memory.staleCandidate);
  const shouldMaskMemory = (memory) =>
    maskableReasons.some((maskableReason) => memory.reasons.includes(maskableReason));
  const maskedDecisions = staleCandidates
    .filter(shouldMaskMemory)
    .map((memory) =>
      freezeDeep({
        memoryId: memory.memoryId,
        memoryKind: normalizeYoungGenerationMemoryKind(memory.metadata),
        staleScore: memory.staleScore,
        retentionValue: memory.retentionValue,
        recencyMs: memory.recencyMs,
        reasons: memory.reasons,
        breakdown: memory.breakdown,
        metadata: memory.metadata,
        inactiveForRetrieval: true,
        masking: freezeDeep({
          isMasked: true,
          maskedAt,
          unmaskedAt: null,
          maskUpdatedAt: maskedAt,
          maskedBy,
          reason,
          maskedOriginalContent: null,
          audit: freezeDeep({
            auditRecordId: provenance?.auditRecordId ?? null,
            policyVersion: provenance?.policyVersion ?? null,
            runtimePhase: provenance?.runtimePhase ?? null,
            sourceEvaluationAt,
            sourceEvaluationMode,
            recordedAt: maskedAt,
            actor: maskedBy,
            metadata: provenance ? freezeDeep(cloneObject(provenance)) : null,
          }),
          provenance: freezeDeep({
            source: "stale-memory-masking-stage",
            sourceEvaluationAt,
            sourceEvaluationMode,
            ...(provenance ?? {}),
          }),
        }),
      }),
    );
  const deferredCandidates = staleCandidates.filter(
    (memory) => !shouldMaskMemory(memory),
  );

  return freezeDeep({
    sourceEvaluationAt,
    sourceEvaluationMode,
    maskedAt,
    maskedBy,
    reason,
    maskableReasons,
    maskedDecisions,
    maskedDecisionCount: maskedDecisions.length,
    deferredCandidates,
    deferredCandidateCount: deferredCandidates.length,
    decisionMode: "offline-suggestion-only",
  });
};
